sale_rental/models/sale_order.py

```diff
@@ -115,37 +145,65 @@ def _prepare_rental(self):
-        res = super(SaleOrderLine, self)._prepare_order_line_procurement(
-            group_id=group_id)
-        if (
-                self.product_id.rented_product_id and
-                self.rental_type == 'new_rental'):
-            # 'product_qty' is re-written after
-            # _prepare_order_line_procurement() in _action_procurement_create()
-            # So I add a key 'rental_product_qty' which is used in the
-            # inherit of create() of procurement.order
-            res.update({
-                'product_id': self.product_id.rented_product_id.id,
-                'product_qty': self.rental_qty,
-                'rental_product_qty': self.rental_qty,
-                'product_uom': self.product_id.rented_product_id.uom_id.id,
-                'location_id':
-                self.order_id.warehouse_id.rental_out_location_id.id,
-                'route_ids':
-                [(6, 0, [self.order_id.warehouse_id.rental_route_id.id])],
-                'date_planned': self._get_rental_date_planned(),
+    def _action_launch_procurement_rule(self):
+        res = super(SaleOrderLine, self)._action_launch_procurement_rule()
+        errors = []
+        for line in self:
+            if not line.product_id.rented_product_id or\
+                    line.rental_type == 'rental_extension':
+                continue
+
+            procurement_group = line.order_id.procurement_group_id
+            if not procurement_group:
+                procurement_group = self.env['procurement.group'].create({
+                    'name': line.order_id.name,
+                    'move_type': line.order_id.picking_policy,
+                    'sale_id': line.order_id.id,
+                    'partner_id': line.order_id.partner_shipping_id.id,
                 })
-        elif self.sell_rental_id:
-            res['route_ids'] = [(6, 0, [
-                self.order_id.warehouse_id.sell_rented_product_route_id.id])]
+                line.order_id.procurement_group_id = procurement_group
+
+            if line.rental_type == 'new_rental':
+                vals = {
+                    'company_id': line.order_id.company_id,
+                    'group_id': procurement_group,
+                    'sale_line_id': line.id,
+                    'date_planned': line._get_rental_date_planned(),
+                    'route_ids': line.order_id.warehouse_id.rental_route_id,
+                    'warehouse_id': line.order_id.warehouse_id or False,
+                    'partner_dest_id': line.order_id.partner_shipping_id
+                }
+            try:
+                self.env['procurement.group'].run(
+                    line.product_id.rented_product_id, line.rental_qty,
+                    line.product_id.rented_product_id.uom_id,
+                    line.order_id.warehouse_id.rental_out_location_id,
+                    line.name, line.order_id.name, vals)
+            except UserError as error:
+                errors.append(error.name)
```

Comment:
I think this try-except-clause should be move into the if-clause. Because "vals" will not be set, if the rental_type is not "new_rental".